File: Transformer.R
Transformer <- torch::nn_module(
  name='Transformer',
  initialize = function(catFeatures, numFeatures=0, numBlocks, dimToken, dimOut=1, 
                        numHeads, attDropout, ffnDropout, resDropout, 
                        headActivation=torch::nn_relu,
                        activation=NULL,
                        ffnNorm=torch::nn_layer_norm, 
                        headNorm=torch::nn_layer_norm,
                        attNorm=torch::nn_layer_norm,
                        dimHidden){
    activation = nn_reglu
    self$Categoricalembedding <- Embedding(catFeatures + 1, dimToken) # + 1 for padding idx
    # self$numericalEmbedding <- numericalEmbedding(numFeatures, dimToken)
    self$classToken <- ClassToken(dimToken)
    
    self$layers <- torch::nn_module_list(lapply(1:numBlocks,
                                                function(x) {
                                                  layer <- torch::nn_module_list()
                                                  layer$add_module('attention', torch::nn_multihead_attention(dimToken,numHeads,
                                                                                                              dropout=attDropout,
                                                                                                              bias=TRUE))
                                                  layer$add_module('ffn', FeedForwardBlock(dimToken, dimHidden,
                                                                                           biasFirst=TRUE,
                                                                                           biasSecond=TRUE,
                                                                                           dropout=ffnDropout,
                                                                                           activation=activation))
                                                  layer$add_module('attentionResDropout', torch::nn_dropout(resDropout))      
                                                  layer$add_module('ffnResDropout', torch::nn_dropout(resDropout))
                                                  layer$add_module('ffnNorm', ffnNorm(dimToken))
                                                  
                                                  if (x!=1) {
                                                    layer$add_module('attentionNorm', attNorm(dimToken))
                                                  }
                                                  return(layer) 
                                                }))
    self$head <- Head(dimToken, bias=TRUE, activation=headActivation, 
                      headNorm, dimOut)
  },
  forward = function(x){
    mask <- torch::torch_where(x ==0, TRUE, FALSE)
    input <- x
    num <- NULL
    cat <- self$Categoricalembedding(x)
    if (!is.null(num)) {
      num <- self$numericalEmbedding(input$num)
      x <- torch::torch_cat(list(cat, num), dim=2L)
      mask <- torch::torch_cat(list(mask, torch::torch_zeros(c(x$shape[1], 
                                                               num$shape[2]), 
                                                             device=mask$device,
                                                             dtype=mask$dtype)), 
                               dim=2L)
    } else {
      x <- cat
    }
    x <- self$classToken(x)
    mask <- torch::torch_cat(list(mask, torch::torch_zeros(c(x$shape[1], 1), 
                                                           device=mask$device,
                                                           dtype=mask$dtype)), 
                             dim=2L)
    for (i in 1:length(self$layers)) {
      layer <- self$layers[[i]]
      xResidual <- self$startResidual(layer, 'attention', x)
      
      if (i==length(self$layers)) {
        dims <- xResidual$shape
        # in final layer take only attention on CLS token
        xResidual <- layer$attention(xResidual[,-1]$view(c(dims[1], 1, dims[3]))$transpose(1,2), 
                                     xResidual$transpose(1,2), 
                                     xResidual$transpose(1,2), mask)
        attnWeights <- xResidual[[2]]
        xResidual <- xResidual[[1]]
        x <- x[,-1]$view(c(dims[1], 1, dims[3]))
      } else {
        # attention input is seq_length x batch_size x embedding_dim
        xResidual <- layer$attention(xResidual$transpose(1,2), 
                                     xResidual$transpose(1,2), 
                                     xResidual$transpose(1,2),
                                     mask,
        )[[1]]
      }
      x <- self$endResidual(layer, 'attention', x, xResidual$transpose(1,2))
      
      xResidual <- self$startResidual(layer, 'ffn', x)
      xResidual <- layer$ffn(xResidual)
      x <- self$endResidual(layer, 'ffn', x, xResidual)
    }
    x <- self$head(x)[,1] # remove singleton dimension
    return(x)
  },
  startResidual = function(layer, stage, x) {
    xResidual <- x
    normKey <- paste0(stage, 'Norm')
    if (normKey %in% names(as.list(layer))) {
      xResidual <- layer[[normKey]](xResidual)
    }
    return(xResidual)
  },
  endResidual = function(layer, stage, x, xResidual) {
    dropoutKey <- paste0(stage, 'ResDropout')
    xResidual <-layer[[dropoutKey]](xResidual)
    x <- x + xResidual
    return(x)
  }
)


FeedForwardBlock <- torch::nn_module(
  name='FeedForwardBlock',
  initialize = function(dimToken, dimHidden, biasFirst, biasSecond,
                        dropout, activation) {
    self$linearFirst <- torch::nn_linear(dimToken, dimHidden*2, biasFirst)
    self$activation <- activation()
    self$dropout <- torch::nn_dropout(dropout)
    self$linearSecond <- torch::nn_linear(dimHidden, dimToken, biasSecond)
  },
  forward = function(x) {
    x <- self$linearFirst(x)
    x <- self$activation(x)
    x <- self$dropout(x)
    x <- self$linearSecond(x)
    return(x)
  }
)

Head <- torch::nn_module(
  name='Head',
  initialize = function(dimIn, bias, activation, normalization, dimOut) {
    self$normalization <- normalization(dimIn)
    self$activation <- activation()
    self$linear <- torch::nn_linear(dimIn,dimOut, bias)
  },
  forward = function(x) {
    x <- x[,-1] # ?
    x <- self$normalization(x)
    x <- self$activation(x)
    x <- self$linear(x)
    return(x)
  }
)

Embedding <- torch::nn_module(
  name='Embedding',
  initialize = function(numEmbeddings, embeddingDim) {
    self$embedding <- torch::nn_embedding(numEmbeddings, embeddingDim, padding_idx = 1)
  },
  forward = function(x_cat) {
    x <- self$embedding(x_cat + 1L) # padding idx is 1L
    return(x)
  }
)

numericalEmbedding <- torch::nn_module(
  name='numericalEmbedding',
  initialize = function(numEmbeddings, embeddingDim, bias=TRUE) {
    self$weight <- torch::nn_parameter(torch::torch_empty(numEmbeddings,embeddingDim))
    if (bias) {
      self$bias <- torch::nn_parameter(torch::torch_empty(numEmbeddings, embeddingDim)) 
    } else {
      self$bias <- NULL
    }
    
    for (parameter in list(self$weight, self$bias)) {
      if (!is.null(parameter)) {
        torch::nn_init_kaiming_uniform_(parameter, a=sqrt(5)) 
      }
    }
  },
  forward = function(x) {
    x <- self$weight$unsqueeze(1) * x$unsqueeze(-1)
    if (!is.null(self$bias)) {
      x <- x + self$bias$unsqueeze(1)
    }
    return(x)
  }
  
)

# adds a class token embedding to embeddings
ClassToken <- torch::nn_module(
  name='ClassToken',
  initialize = function(dimToken) {
    self$weight <- torch::nn_parameter(torch::torch_empty(dimToken,1))
    torch::nn_init_kaiming_uniform_(self$weight, a=sqrt(5))
  },
  expand = function(dims) {
    newDims <- vector("integer", length(dims) - 1) + 1
    return (self$weight$view(c(newDims,-1))$expand(c(dims, -1)))
    
  },
  forward = function(x) {
    return(torch::torch_cat(c(x, self$expand(c(dim(x)[[1]], 1))), dim=2))
  }
)

nn_reglu <- torch::nn_module(
  name='ReGlu',
  forward = function(x) {
    return(reglu(x))
  }
)


reglu <- function(x) {
  chunks <- x$chunk(2, dim=-1)
  
  return(chunks[[1]]* torch::nnf_relu(chunks[[2]]))
  
}

File: Transformer.py
import math
import torch
from torch import nn
import torch.nn.functional as F


class ReGLU(nn.Module):
    def forward(self, x):
        a, b = x.chunk(2, dim=-1)
        return a * F.relu(b)


class Transformer(nn.Module):
    def __init__(self, catFeatures=None, numBlocks=2, dimToken=64, dimout=1,
                 numHeads=8, attDropout=0.2, ffnDropout=0.2, resDropout=0,
                 headActivation=nn.ReLU,
                 activation=ReGLU,
                 ffnNorm = nn.LayerNorm,
                 headNorm = nn.LayerNorm,
                 attNorm = nn.LayerNorm,
                 dimHidden=512):
        super(Transformer, self).__init__()
        self.CategoricalEmbedding = Embedding(catFeatures + 1, dimToken)
        self.ClassToken = ClassToken(dimToken)
        self.layers = nn.ModuleList([])
        for layer_idx in range(numBlocks):
            layer = nn.ModuleDict({
                'attention': nn.MultiheadAttention(dimToken, numHeads,
                                                   dropout=attDropout, bias=True,
                                                   batch_first=True),
                'ffn': FeedForwardBlock(dimToken, dimHidden,
                                        biasFirst=True, biasSecond=True, dropout=ffnDropout,
                                        activation=activation),
                'attentionResDropout': nn.Dropout(resDropout),
                'ffnResDropout': nn.Dropout(resDropout),
                'ffnNorm': ffnNorm(dimToken)
            })
            if layer_idx != 0:
                layer['attentionNorm'] = attNorm(dimToken)
            self.layers.append(layer)

        self.head = Head(dimToken, bias=True, activation=headActivation,
                         norm=headNorm, dimOut=dimout)

    def forward(self,input):
        mask = torch.where(input == 0, True, False)
        x = self.CategoricalEmbedding(input)
        x = self.ClassToken(x)
        mask = torch.cat([mask, torch.zeros((x.shape[0], 1), device=mask.device,
                                           dtype=mask.dtype)], dim=1)
        for layer_idx, layer in enumerate(self.layers):
            xResidual = self.startResidual(layer, 'attention', x)

            if layer_idx==len(layer):
                dims = xResidual.shape
                xResidual, attnWeights = layer['attention'](
                    xResidual[:,-1].view((dims[0], 0, dims[2])),
                    xResidual,
                    xResidual, mask
                )
                x = x[:,-1].view(dims[0], 0, dims[2])
            else:
                xResidual, _ = layer['attention'](xResidual,
                                                  xResidual,
                                                  xResidual,
                                                  mask)
            x = self.endResidual(layer, 'attention', x, xResidual)

            xResidual = self.startResidual(layer, 'ffn', x)
            xResidual = layer['ffn'](xResidual)

            x = self.endResidual(layer, 'ffn', x, xResidual)

        x = self.head(x).squeeze()
        return x

    def startResidual(self, layer, stage, x):
        xResidual = x
        normKey = f'{stage}Norm'
        if normKey in layer:
            xResidual = layer[normKey](xResidual)
        return xResidual

    def endResidual(self, layer, stage, x, xResidual):
        xResidual = layer[f'{stage}ResDropout'](xResidual)
        x = x + xResidual
        return x


class FeedForwardBlock(nn.Module):
    def __init__(self, dimToken=64, dimHidden=512, biasFirst=True, biasSecond=True,
                 dropout=0.0, activation=nn.ReLU):
        super(FeedForwardBlock, self).__init__()
        self.linear_first = nn.Linear(dimToken, dimHidden * 2, biasFirst)
        self.activation = activation()
        self.dropout = nn.Dropout(dropout)
        self.linear_second = nn.Linear(dimHidden, dimToken, biasSecond)

    def forward(self, input):
        x = self.linear_first(input)
        x = self.activation(x)
        x = self.dropout(x)
        x = self.linear_second(x)
        return x


class Head(nn.Module):

    def __init__(self, dim_in, bias, activation, norm, dimOut):
        super(Head, self).__init__()
        self.normalization = norm(dim_in)
        self.activation = activation()
        self.linear = nn.Linear(dim_in, dimOut, bias)

    def forward(self, input):
        x = input[:, -1]  # ?
        x = self.normalization(x)
        x = self.activation(x)
        x = self.linear(x)
        return x


class Embedding(nn.Module):

    def __init__(self, numEmbeddings, embeddingDim):
        super(Embedding, self).__init__()
        self.embedding = nn.Embedding(numEmbeddings, embeddingDim, padding_idx=1)

    def forward(self, input):
        x = self.embedding(input + 1)
        return x


# adds a class token embedding
class ClassToken(nn.Module):
    def __init__(self, dimToken):
        super(ClassToken, self).__init__()
        self.weight = nn.Parameter(torch.Tensor(dimToken, 1))

        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))

    def expand(self,*dims):
        if not dims:
            return self.weight
        new_dims = (1,) * (len(dims) -1)
        return self.weight.view(*new_dims, -1).expand(*dims, -1)

    def forward(self, input):
        return torch.cat([input, self.expand(len(input), 1)], dim=1)

